Migrate tests to elm-check 3

jivagoalves opened the pull request:
After migrating the tests to latest elm-check (3), I've found a failure related to the float parser when dealing with 0. I suspect we need to use `0.0` instead of `0` when generating the floats for the property test but wanted to check with you first.

![screenshot from 2015-06-05 23 04 06](https://cloud.githubusercontent.com/assets/887937/8018261/ac9f75f6-0bd7-11e5-88cd-fa6c4e62a904.png)

Would you mind helping me out with this issue?

Besides that, I had to download [Shrink.elm](https://raw.githubusercontent.com/TheSeamau5/elm-shrink/master/src/Shrink.elm) to be able to compile even though it's already downloaded as a dependency. I suspect the problem is because I'm on Linux (case-sensitive FS) and [Shrink.Elm has been renamed recently to Shrink.elm](https://github.com/TheSeamau5/elm-shrink/commit/f6ecd39751eea2c3e1952b43cfa7be3bea31c93a). They didn't release this change yet. So if you have problems compiling this PR because of "Shrink module not found", then copying this might help.

Thanks!

Dandandan commented:
Thanks, this is great! I stumbled into the "Shrink module not found" error before, didn't know it was because of that (also on a Linux machine).

I think we want restrict to the `0.0` format, so we should change the test.

jivagoalves commented:
@Dandandan I pushed a commit on top of the PR (for easier review) to use `0.0` instead. We can squash them when we're done with the PR.

Now I'm experiencing this error below:
![screenshot from 2015-06-06 17 11 52](https://cloud.githubusercontent.com/assets/887937/8021689/2a4c5f18-0c6f-11e5-8427-9378ccb5d1b6.png)

It is a floating-point error. I did some tweaks in the float parsing algorithm but it's still breaking for other cases. I'm looking at [this](http://docs.oracle.com/cd/E19957-01/806-3568/ncg_goldberg.html) to brush up my skills on floating-point arithmetic. Please let me know if you have any idea to help fixing this one.

Dandandan commented:
Thanks again

1. We may want to change the test to something like

`let s = toString n in if String.contains "."  s` then s else s + ".0"

to also pass tests when generating 2, 3 etc.

2.

Hm that's interesting. What about converting the list to a `String` and using `String.toFloat`?

jivagoalves commented on Test.elm:
```diff
@@ -1,64 +1,83 @@
+        (parse PN.float << (\n -> let s = toString n
+                                  in if String.contains "." s
+                                        then s
+                                        else s ++ ".0"))
```
I changed the test to append `.0`.

jivagoalves commented on Parser/Number.elm:
```diff
@@ -34,9 +35,17 @@ integer : Parser Int
+    let toFloatString (i,ds) =
+          toString i ++ List.foldl (\d s -> s ++ toString d) "." ds
+        convertToFloat sig int digs =
+          toFloat sig * (fromOk << String.toFloat << toFloatString) (int,digs)
```
@Dandandan I followed your suggestion and now I'm converting to string and then using `String.toFloat`. Since we have a `Result e a`, we need to extract out the `a` from Ok. OTOH, there's the `e` case. I believe the parser will give us a valid float and at this point we shouldn't worry about that case. I think the reasonable approach would be to fail somehow instead of providing a default value if we see this exception (in theory shouldn't happen). What do you think about the use of unsafe `fromOk` in this particular case? Is there anything else we could improve? :)

Dandandan commented:
I created a pull request against your repository!

jivagoalves commented on elm-package.json:
```diff
@@ -12,8 +12,12 @@
     ],
     "dependencies": {
-        "elm-lang/core": "2.0.1 <= v < 2.0.2",
+        "TheSeamau5/elm-check": "3.2.0 <= v < 4.0.0",
```
Bumped to `3.2.0` which now has [lowerCaseChar and upperCaseChar](https://github.com/TheSeamau5/elm-check/blob/master/src/Check/Investigator.elm#L143-L155).

jivagoalves commented on Parser/Number.elm:
```diff
@@ -32,11 +35,18 @@ sign =
 float =
-    let convertToFloat sig int dig = toFloat sig * (toFloat int + 0.1 * List.foldr (\b a -> a / 10 + b) 0 (List.map toFloat dig))
+    let toFloatString (i,ds) =
+          toString i ++ "." ++ String.concat (List.map toString ds)
```
looks easier to read

jivagoalves commented on Parser/Number.elm:
```diff
@@ -32,11 +35,18 @@ sign =
+
+{-| The fromOk function extracts the element out of a Ok, with a default. -}
+fromOk : a -> Result e a -> a
+fromOk d = withDefault d << toMaybe
```
I'm relying upon core functions here instead.

jivagoalves commented:
@Dandandan I pushed some small changes: bumped elm-check to get latest investigators and `fromOk` to rely upon core functions. Please let me know if I can squash this so that we cam merge. Thanks again!

Dandandan commented:
Looks great! It seems we can remove the dependency "elm-random-extra" now. Maybe you could check this, then let's merge it!

jivagoalves commented:
I removed the dependency and squashed the commits. Hopefully this PR should be good to merge now. Please let me know if there's anything else I can improve. Thanks for your help!

Dandandan commented:
Thanks again @jivagoalves for all the work!
